### PR TITLE
Updated `inPolygon`

### PR DESCRIPTION
**Pull request by Chillee**
Built on top of #78 , so review/merge that one first. I'll rebase when that's merged.

This is currently mostly just for keeping track of what I'm working on.

### CODE REVIEW

File: content/geometry/insidePolygon.h [hunk 1]
```diff
@@ -1,34 +1,29 @@
...
-		n += (max(i->y,j->y) > p.y && min(i->y,j->y) <= p.y &&
-				((*j-*i).cross(p-*i) > 0) == (i->y <= p.y));
-	}
-	return n&1; //inside if odd number of intersections
+template<class P>
+bool inPolygon(vector<P> &p, P a, bool strict = true) {
+		int nCross = 0, n = sz(p);
+		rep(i, 0, n) {
+			if (onSegment(p[i], p[(i + 1) % n], a)) return !strict;
+			nCross += (sgn(a.y >= p[(i + 1) % n].y) -
```
Comment (simonlindholm) [inline content/geometry/insidePolygon.h]: sgn(bool) looks off, was this meant as sgn(everything) > 0?

Comment (Chillee) [reply content/geometry/insidePolygon.h]: You're right that there should probably be a `sgn` around `a.cross`. Take a look at page 61 here for an explanation of the method: https://vlecomte.github.io/cp-geo.pdf

Comment (Chillee) [reply content/geometry/insidePolygon.h]: I think I updated it - should resolve issues.

**Comment by Chillee** [comment]
The logic for checking whether it crosses the line is fundamentally the same, so I don't think the behavior for numerical accuracy should be better/worse. It's mainly just a shortened version (with the observation that subtracting the booleans should result in the same sign as the cross product if it intersects).

image

**Comment by Chillee** [comment]
Kattis PointInPolygon submission: https://ideone.com/T4D6qk

Couldn't find `copsrobbers` or `intersection`.

File: content/geometry/onSegment.h [hunk 2]
```diff
@@ -4,6 +4,7 @@
  * License: CC0
  * Source: https://vlecomte.github.io/cp-geo.pdf
  * Description: Returns true iff p lies on the line segment from s to e. Intended for use with e.g. Point<ll> where overflow is an issue.  Use \texttt{(segDist(s,e,p)<=epsilon)} instead when using Point<double>.
+ * Works with integers.
```
Comment (simonlindholm) [inline content/geometry/onSegment.h]: There's already a comment about about integers/doubles; I don't think this adds anything. (That comment could be reformulated to drop the "e.g." and "where overflow is an issue"; `Point<int>` should die)

File: content/geometry/insidePolygon.h [hunk 3]
```diff
@@ -1,34 +1,31 @@
...
-	}
-	return n&1; //inside if odd number of intersections
+template<class P>
+bool inPolygon(vector<P> &p, P a, bool strict = true) {
+		int cnt = 0, n = sz(p);
+		rep(i,0,n) {
+			P np = p[(i + 1) % n];
+			if (onSegment(p[i], np, a)) return !strict;
+			//or: if (segDist(p[i], np, a) <= eps) return !strict;
+			cnt += ((a.y>=np.y)-(a.y>=p[i].y))*a.cross(p[i],np) > 0;
```
Comment (simonlindholm) [inline content/geometry/insidePolygon.h]: `cnt ^=`, to get rid of the bitand at the end

File: content/geometry/insidePolygon.h [hunk 4]
```diff
@@ -1,34 +1,31 @@
...
-		//increment n if segment intersects line from p
-		n += (max(i->y,j->y) > p.y && min(i->y,j->y) <= p.y &&
-				((*j-*i).cross(p-*i) > 0) == (i->y <= p.y));
-	}
-	return n&1; //inside if odd number of intersections
+template<class P>
+bool inPolygon(vector<P> &p, P a, bool strict = true) {
+		int cnt = 0, n = sz(p);
+		rep(i,0,n) {
+			P np = p[(i + 1) % n];
```
Comment (simonlindholm) [inline content/geometry/insidePolygon.h]: we could do `np = p[(i ?: n) - 1]` to get rid of the modulo

File: content/geometry/insidePolygon.h (hunk 3 above)
Comment (simonlindholm) [reply content/geometry/insidePolygon.h]: also, I'd appreciate some more whitespace here (e.g. around the `*`), I think it's hard to follow where parentheses end. I think you could win some characters by converting `>=` to `<`? and by renaming `np` to `q`, if needed

File: content/geometry/insidePolygon.h [hunk 5]
```diff
@@ -1,34 +1,31 @@
 /**
- * Author: Ulf Lundstrom
- * Date: 2009-03-22
+ * Author: Victor Lecomte, chilli
+ * Date: 2019-04-26
  * License: CC0
- * Source: Basic geometry
- * Description: Returns true if p lies within the polygon described by the points between iterators begin and end. If strict false is returned when p is on the edge of the polygon. Answer is calculated by counting the number of intersections between the polygon and a line going from p to infinity in the positive x-direction. The algorithm uses products in intermediate steps so watch out for overflow. If points within epsilon from an edge should be considered as on the edge replace the line "if (onSegment..." with the comment bellow it (this will cause overflow for int and long long).
+ * Source: https://vlecomte.github.io/cp-geo.pdf
+ * Description: Returns true if p lies within the polygon. If p is on the edge of the polygon, it returns false when strict is true. Answer is calculated by counting the number of intersections between the polygon and a line going from p to infinity in the positive x-direction. The algorithm uses products in intermediate steps so watch out for overflow. Works with integers.
```
Comment (simonlindholm) [inline content/geometry/insidePolygon.h]: again, the `Works with integers.` comment is pretty much implicit. The interesting question is whether it works with doubles. It seems to have some special-casing for it, but I'm not quite sure about the handling of parallel lines, I'd need to think about it.

File: content/geometry/insidePolygon.h [hunk 6]
```diff
@@ -1,34 +1,31 @@
...
+ * Status: fuzz-tested and Kattis problem pointinpolygon
  * Usage:
- * 	typedef Point<int> pi;
- * 	vector<pi> v; v.push_back(pi(4,4));
- * 	v.push_back(pi(1,2)); v.push_back(pi(2,1));
- * 	bool in = insidePolygon(v.begin(),v.end(), pi(3,4), false);
+ * 	typedef Point<int> P;
+ * 	vector<P> v; v.push_back(P(4,4));
+ * 	v.push_back(P(1,2)); v.push_back(P(2,1));
+ * 	bool in = insidePolygon(v, P(3,4), false);
```
Comment (simonlindholm) [inline content/geometry/insidePolygon.h]: does the following work?
```cpp
vector<Point<ll>> v = {{2,1}, {1,2}, {2,1}};
bool in = inPolygon(v, {3, 4}, false);
```
Either way, `int`-> `ll`, insidePolygon -> inPolygon

File: content/geometry/insidePolygon.h (hunk 5 above)
Comment (simonlindholm) [reply content/geometry/insidePolygon.h]: Also, can you line-wrap this description?

File: content/geometry/insidePolygon.h (hunk 4 above)
Comment (Chillee) [reply content/geometry/insidePolygon.h]: I did not know that was acceptable syntax.

File: content/geometry/insidePolygon.h (hunk 6 above)
Comment (Chillee) [reply content/geometry/insidePolygon.h]: I don't *really* understand the example already existing or the example you posted. I think it's better to have the function return 1.

I'll do
```
    vector<Point<ll>> v= {P{4,4}, P{1,2}, P{2,1}};
    bool in = inPolygon(v, P{3, 3}, false);
```

File: content/geometry/onSegment.h (hunk 2 above)
Comment (Chillee) [reply content/geometry/onSegment.h]: Dropping those parts results in `Intended for use with Point<ll>`. I think `works with integers` is a clearer way of expressing the same thing.

File: fuzz-tests/utilities/genPolygon.h [hunk 7]
```diff
@@ -0,0 +1,70 @@
+#pragma once
+
+// Translated from Python code posted here: https://codeforces.com/blog/entry/63058?#comment-472788
```
[the remark this answers is not in the document]
Comment (Chillee) [reply fuzz-tests/utilities/genPolygon.h]: No to duplicates, yes to colinear.

File: fuzz-tests/utilities/genPolygon.h [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    auto pa = conquer(S[idx0], depth+1);
+    auto pb = conquer(S[!idx0], depth+1);
+    if (!pa.first || !pb.first) return {false, {}};
+    pb.second.erase(pb.second.begin());
+    pa.second.insert(pa.second.end(), all(pb.second));
+    return pa;
+}
+template<class P> vector<P> genPolygon(vector<P> pts) {
+    sort(all(pts));
+    pts.resize(unique(all(pts), [](auto a, auto b){return (a-b).dist()<EPS;}) - pts.begin());
```
[the remark this answers is not in the document]
Comment (Chillee) [reply fuzz-tests/utilities/genPolygon.h]: This is more just to eliminate duplicate points. I'll just use the default comparator.

File: content/geometry/insidePolygon.h (hunk 3 above)
Comment (Chillee) [reply content/geometry/insidePolygon.h]: I added some whitespace.

File: content/geometry/insidePolygon.h (hunk 5 above)
Comment (Chillee) [reply content/geometry/insidePolygon.h]: Line wrapped the description.

Also, why is it implicit? It's obvious if you understand that all the operations are just cross/dot products, but I think it's worth noting so you can understand at a glance.

File: content/geometry/insidePolygon.h (hunk 4 above)
Comment (simonlindholm) [reply content/geometry/insidePolygon.h]: It's a GCC extension, I can totally understand if you dislike it...

File: content/geometry/insidePolygon.h (hunk 5 above)
Comment (simonlindholm) [reply content/geometry/insidePolygon.h]: All algorithms work with integers if they don't have floating point results, and if it didn't work with integers, it would do `typedef Point<double> P;`.

Comment (Chillee) [reply content/geometry/insidePolygon.h]: I wasn't sure that was the intended standard, since I saw that `circleTangents` was written like this:
```cpp
template<class P>
pair<P,P> circleTangents(const P &p, const P &c, double r) {
}
```
but had this in the description:

" P is intended to be Point<double>"

File: content/geometry/insidePolygon.h (hunk 4 above)
Comment (Chillee) [reply content/geometry/insidePolygon.h]: I think I'd rather not - I've never used it before :^)

File: content/geometry/insidePolygon.h (hunk 5 above)
Comment (simonlindholm) [reply content/geometry/insidePolygon.h]: Oh, we should just update circleTangents in that case. `typedef` is clearer than a description note, I think.

File: content/geometry/insidePolygon.h (hunk 4 above)
Comment (simonlindholm) [reply content/geometry/insidePolygon.h]: It can allow for some nice golfing, see e.g. the SCC. But yeah, let's leave it I guess.

File: content/geometry/onSegment.h (hunk 2 above)
Comment (simonlindholm) [reply content/geometry/onSegment.h]: I removed even the "Works with integers." -- I think it's pretty implicit from how most geometry algorithms operate on ints, and how the description says how to modify it for doubles.
